### ISSUE TITLE
Clang doesn't emit `@llvm.is.constant.i32` for function calls in `__builtin_constant_p`

### ISSUE THREAD

**Issue by Eisenwave** [llvm/llvm-project#65741]
```cpp
int foo() { return 37; }

int bar() {
    return __builtin_constant_p(foo());
}
```
The following IR is emitted for `bar` (https://godbolt.org/z/eWW1a9P4v):
```llvm
define dso_local noundef i32 @bar()() {
entry:
  ret i32 0
}
```
I would expect something along the lines of
```llvm
define dso_local noundef i32 @zap(int)(i32 noundef %z) local_unnamed_addr {
entry:
  %call = call noundef i32 @foo()() ; only emitted if foo() is has no side effects
  %0 = tail call i1 @llvm.is.constant.i32(i32 %call)
  %1 = zext i1 %0 to i32
  ret i32 %1
}
```

To me, this looks like a missed optimization. `foo()` can easily be constant folded and the result of `foo()` should be known. If `@llvm.is.constant` was emitted, I'm sure that it would result in `1` as well. How come Clang doesn't emit it at all?

**Comment by Eisenwave** [comment]
Hmm, this might be a duplicate of https://github.com/llvm/llvm-project/issues/42294, but wouldn't it be possible to infer from the definition of `foo` that a call to it doesn't have side effects?

**Comment by Eisenwave** [comment]
Well, putting `[[gnu::const]]` on `foo()` turns it *exactly* into the IR that I've suggested, so this is basically just a skill issue on the side of the user.

**Comment by cor3ntin** [comment]
The front end (?) might decides that rvalues are never constant
https://godbolt.org/z/9T3d5zv4Y

```cpp
void test1() {
    int a = __builtin_constant_p(foo());
}
```

Produces, when compiled with ` -disable-llvm-passes`

```llvm
define dso_local void @test1()() #0 {
  %1 = alloca i32, align 4
  store i32 0, ptr %1, align 4
  ret void
}
```

Compared to the lvalue case 

```cpp
void test2() {
    int i = foo();
    int a = __builtin_constant_p(i);
}
```

```llvm
define dso_local void @test2()() #0 {
  %1 = alloca i32, align 4
  %2 = alloca i32, align 4
  %3 = call noundef i32 @foo()()
  store i32 %3, ptr %1, align 4
  %4 = load i32, ptr %1, align 4
  %5 = call i1 @llvm.is.constant.i32(i32 %4)
  %6 = zext i1 %5 to i32
  store i32 %6, ptr %2, align 4
  ret void
}
```

Which  does emit  `@llvm.is.constant`
The fix might be to emit a warning in the front end when `__builtin_constant_p` is used with an rvalue

**Comment by cor3ntin** [comment]
https://github.com/llvm/llvm-project/blob/feb7beaf70bace1c3ffafc7f732c8fadca5e8c9d/clang/lib/CodeGen/CGBuiltin.cpp#L3048-L3051

We just silently produce `0` when the expression might have side effects.
We probably want a diagnostic there.
Definitively a front end issue then

**Comment by Eisenwave** [comment]
> The fix might be to emit a warning in the front end when `__builtin_constant_p` is used with an rvalue

I don't think it has anything to do with value categories. It's that the expression in `__builtin_constant_p` is not evaluated. If you put `int i = foo();` outside, that's obviously not an obstacle because your program is evaluating `foo`.
